lib/SILOptimizer/Analysis/MemoryBehavior.cpp

```diff
@@ -509,6 +512,34 @@ visitBeginCOWMutationInst(BeginCOWMutationInst *BCMI) {
   return MemBehavior::None;
 }
 
+MemBehavior MemoryBehaviorVisitor::visitLoadBorrowInst(LoadBorrowInst *lbi) {
+  if (!mayAlias(lbi->getOperand()))
+    return MemBehavior::None;
+  return MemBehavior::MayRead;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
+  return MemBehavior::None;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitEndBorrowInst(EndBorrowInst *ebi) {
+  auto borrowIntroducer = ebi->getOperand();
+  // An end_borrow is defined as a MemoryBehaviour::MayRead of the load borrowed
+  // address. This prevents other aliasing memory writes to be moved across the
+  // end_borrow.
+  if (auto *lbi = dyn_cast<LoadBorrowInst>(borrowIntroducer)) {
+    if (!mayAlias(lbi->getOperand()))
+      return MemBehavior::None;
+  }
+  if (auto *bbi = dyn_cast<BeginBorrowInst>(borrowIntroducer)) {
+    return MemBehavior::None;
```

Comment:
Making ``end_borrow`` a non-side-effect instruction means that all optimizations which try to move memory loads down the control flow must now handle end_borrow explicitly - like you did in ``TempRValueOptPass::extendAccessAndBorrowScopes``.
I think it's okay, but we have to be aware of this.